### ISSUE TITLE
[Feature Request] Package Manager package support

### ISSUE THREAD

**Issue by KesKim** [YarnSpinnerTool/YarnSpinner#201]
I've been eyeing and using the Unity's Package Manager system for a bit now, and just learned about Scoped Registries and OpenUPM.

Generally, updating anything via Package Manager has a lot less friction than the old method of browsing, downloading .unitypackage file, opening it in Unity, checking the files its outputting and in general keeping the files in your Assets, part of your compile load each time you edit your code.

Scoped registries seem to already allow describing alternate repositories for packages to display as downloadable/updateable in Package Manager window. https://docs.unity3d.com/2020.1/Documentation/Manual/upm-scoped.html

OpenUPM is one such registry https://openupm.com/

In summary: it would be lovely to have Yarn Unity integration as a Package Manager supported version, possibly via some scoped registry. ❤️

**Comment by desplesda** [comment]
Yeah, this is absolutely something we want to have, particularly since we'll be releasing new versions over time and want an easy way to upgrade.

**Comment by Schroedingers-Cat** [comment]
Adding YarnSpinner to OpenUPM should be straightforward from now on that the changes from #195 have been merged.  
I had a read through the OpenUPM docs and it seems that all that needs to be done is: 
* [ ] (Optional) Changing the directory name `Examples` to `Samples~` on the `development` branch and adding the sample files to the `project.json` like [described here](https://forum.unity.com/threads/samples-in-packages-manual-setup.623080/)
  * This will tell Unity to not import this content until you click on `Import in project` in the Package Manager UI. This is fine for UPM users but users using the .unitypackage or people developing example content will need to manually place these files outside of the `Sample~` directory, preferable under `Assets`. 
  * After developing, developers will also need to manually move the changed files back to `Samples~`.
  * A workaround for the developers (not the .unitypackage users) is to version control everything under the current `Examples` directory but use the CI to change that directory to `Samples~` upon release.
* [ ] Changing the version in the `package.json` to the version of the new release (e.g. `1.1.0`)
* [ ] Merging the `development` branch to the `master` branch so the changes from #195 are available on the `master` branch
* [ ] Creating a new release (which creates a new semantic versioning tag in the git repo) or tagging that commit with a upm targeting tag like `upm/1.1.0`
* [ ] Adding the YarnSpinner repo to their registry via [this form](https://openupm.com/packages/add/) 
* [ ] (Optional) Adding the OpenUPM badge to this repo's README file like [described here](https://openupm.com/docs/adding-badge.html)

Upon registration, the repository is being scanned for a `package.json` file in the file tree of the `master` branch. Once that file is found,

**Comment by favoyang** [comment]
Hi there,

Favo from OpenUPM. Glad to see some processes here to bring UPM support for YarnSpinner. The repository has been submitted to OpenUPM. However, it seems not working because of lacking the DLLs.

OpenUPM build pipelines simply scan new Git tags, then locate the package.json, and bundle a package here. Well, this works for most repositories, but this one requires a custom build process to generate those DLLs. We don't support custom build process yet, because it is a case by case thing. Instead, we encourage developers to use CI to create an always up-to-date and ready-to-use upm branch, then tag a versioned `upm/x.y.z` along with the main release tag `x.y.z`.

I have prepared PR #238 

- Add GitHub action to create upm branch
- DLLs generated within the process
- DLL Unity meta files are stored under `.github/meta`, and copied to the upm branch. This is so far the cheapest way to add those meta files without launching Unity on the CI machine.
- The examples folder has been handled the UPM way (using the Unity package manager to import samples)

You're welcome to test the result with this added to your Packages/manifest.json

```
"dev.secretlab.yarnspinner": "https://github.com/favoyang/YarnSpinner.git#upm"
```

Go to Package Manager UI, locate the package, then import the sample folder to give a try. It works for me.

If this makes sense after a review from the core team member. We can move to the next steps
- Tag `upm/x.y.z` from the upm branch when creating a new release
- Add some documentation: changelog, readme, and badge.

**Comment by desplesda** [comment]
Yarn Spinner for Unity is now available via OpenUPM. Installation information is available via the [Yarn Spinner site](https://yarnspinner.dev/docs/unity/installing/). Closing this issue now; if you need any help, please file an issue on the [Yarn Spinner for Unity repo](https://github.com/YarnSpinnerTool/YarnSpinner-Unity)!
